src/libPMacc/include/dataManagement/DataConnector.hpp

```diff
@@ -154,68 +154,86 @@ namespace PMacc
-        template<class TYPE>
-        TYPE &getData(SimulationDataId id, bool noSync = false)
+        template< class TYPE >
+        std::shared_ptr< TYPE >
+        get(
+            SimulationDataId id,
+            bool noSync = false // @todo invert!
+        )
         {
-            std::map<SimulationDataId, Dataset*>::const_iterator iter = datasets.mapping.find(id);
+            auto it = findId( id );
+
+            if( it == datasets.end() )
+                throw std::runtime_error(
+                    getExceptionStringForID(
+                        "Invalid dataset ID",
+                        id
+                    )
+                );
 
-            if (iter == datasets.mapping.end())
-                throw std::runtime_error(getExceptionStringForID("Invalid DataConnector dataset ID", id));
+            log< ggLog::MEMORY >( "DataConnector: sharing access to '%1%' (%2% uses)" ) %
+                                id % ( it->use_count() );
 
-            Dataset * dataset = iter->second;
-            if (!noSync)
+            if( !noSync )
             {
-                dataset->synchronize();
+                (*it)->synchronize();
             }
 
-            return (TYPE&) (dataset->getData());
+            return std::static_pointer_cast< TYPE >( *it );
         }
 
-        /**
-         * Decrements the reference counter to the data specified by id.
+        /** Indicate a data set gotten temporarily via @see getData is not used anymore
+         *
+         * @todo not implemented
          *
          * @param id id for the dataset previously acquired using getData()
          */
-        void releaseData(SimulationDataId)
+        void
+        releaseData( SimulationDataId )
         {
         }
 
     private:
 
-        friend class Environment<DIM1>;
-        friend class Environment<DIM2>;
-        friend class Environment<DIM3>;
+        friend class Environment< DIM1 >;
+        friend class Environment< DIM2 >;
+        friend class Environment< DIM3 >;
 
-        static DataConnector& getInstance()
+        static DataConnector&
+        getInstance()
         {
             static DataConnector instance;
             return instance;
         }
 
-        Mapping<SimulationDataId, Dataset*> datasets;
+        std::list< std::shared_ptr< ISimulationData > > datasets;
 
         DataConnector()
         {
-            datasets.sorter = new ListSorter<SimulationDataId > ();
         };
 
-        virtual ~DataConnector()
+        virtual
+        ~DataConnector()
         {
-            std::map<SimulationDataId, Dataset*>::const_iterator iter;
-            for (iter = datasets.mapping.begin(); iter != datasets.mapping.end(); iter++)
-                delete iter->second;
-
-            if (datasets.sorter != nullptr)
+            log< ggLog::MEMORY >( "DataConnector: being destroyed (%1% datasets left to destroy)" ) %
+                                datasets.size();
+            //datasets.clear();
```

Comment:
@n01r little tear-down bug here that I discovered yesterday.
use this line and remove the for loop or change the for loop for now (RNG seems to forget to unshare itself)